lucet-runtime/lucet-runtime-internals/src/region/uffd.rs

```diff
@@ -0,0 +1,512 @@
+        if alloc.heap_accessible_size > 0 {
+            unsafe {
+                madvise(
+                    alloc.slot().heap,
+                    alloc.heap_accessible_size,
+                    MmapAdvise::MADV_DONTNEED,
+                )?;
+            }
+        }
+
+        // reset the heap to the initial size
+        let initial_size = module
+            .heap_spec()
+            .map(|h| h.initial_size as usize)
+            .unwrap_or(0);
+        alloc.heap_accessible_size = initial_size;
+        alloc.heap_inaccessible_size = alloc.slot().limits.heap_address_space_size - initial_size;
+        Ok(())
+    }
+
+    fn get_limits(&self) -> &Limits {
+        &self.limits
+    }
+
+    fn as_dyn_internal(&self) -> &dyn RegionInternal {
+        self
+    }
+}
+
+impl RegionCreate for UffdRegion {
+    const TYPE_NAME: &'static str = "UffdRegion";
+
+    fn create(instance_capacity: usize, limits: &Limits) -> Result<Arc<Self>, Error> {
+        UffdRegion::create(instance_capacity, limits)
+    }
+}
+
+impl Drop for UffdRegion {
+    fn drop(&mut self) {
+        // eprintln!("UffdRegion::drop()");
+        // write to the pipe to notify the handler to exit
+        if let Err(e) = nix::unistd::write(self.handler_pipe, b"macht nichts") {
+            // this probably means the handler errored out; note it but don't panic
+            eprintln!("couldn't write to handler shutdown pipe: {}", e);
+        };
+
+        // eprintln!("joining");
+        // wait for the handler to exit
+        let res = self
+            .handler
+            .take()
+            .expect("region has a join handle")
+            .join()
+            .expect("join on uffd handler");
+
+        // close the send end of the pipe; the handler closes the other end
+        nix::unistd::close(self.handler_pipe).expect("close handler exit pipe");
+
+        let total_region_size = self.instance_capacity * self.limits.total_memory_size();
+        unsafe {
+            munmap(self.start, total_region_size).expect("unmapping region");
+        }
+
+        if let Err(e) = res {
+            panic!("uffd handler thread failed: {}", e);
+        }
+    }
+}
+
+impl UffdRegion {
+    pub fn create(instance_capacity: usize, limits: &Limits) -> Result<Arc<Self>, Error> {
+        assert!(
+            limits.signal_stack_size % host_page_size() == 0,
+            "signal stack size is a multiple of host page size"
+        );
+        if instance_capacity == 0 {
+            return Err(Error::InvalidArgument(
+                "region must be able to hold at least one instance",
+            ));
+        }
+        limits.validate()?;
+
+        let uffd = Arc::new(
+            UffdBuilder::new()
+                .close_on_exec(true)
+                .non_blocking(true)
+                .create()
+                .map_err(|e| Error::InternalError(e.into()))?,
+        );
+
+        // map the chunk of virtual memory for all of the slots
```

Comment:
I'm guessing the answer to my question could be here.  The memory is already marked as READ/WRITE for all the slots.

Reply:
https://github.com/bytecodealliance/lucet/pull/492/files#diff-d382c16395ef5840f0796f78cf91450dR136-R139 is also a key part of why this works

Reply:
Hm.  If I'm reading that correctly, does that mean that the stack isn't zeroed out until it's actually needed?

Reply:
It may help to refer to the `ioctl_userfaultfd(2)` man page on this, since "zeropage" does a lot more than what it sounds like: http://man7.org/linux/man-pages/man2/ioctl_userfaultfd.2.html (search for `UFFDIO_ZEROPAGE` on that page).

Basically `zeropage` is one of several ways that the uffd worker thread can resolve a page fault in the memory that it has registered. The stack starts off as `mprotect` read/write but `madvise` don't need, so accesses will trigger a uffd fault. We then recognize that the fault address is in the stack, and then call `zeropage` to both zero out the memory and resolve the fault, so that the instance's thread wakes up and can make progress once again.